### ISSUE TITLE
Set comprehension should verify element-type subset constraints

### ISSUE THREAD

**Issue by RustanLeino** [dafny-lang/dafny#697]
A subset type

    type T = x: Base | P(x)

denotes the subset of values of type `Base` that satisfy `P`. The constraint `P(x)` is allowed to be a ghost expression and does need to be compilable. Now, here's the problem:

```
datatype Cell = Cell(x: int)
type EvenCell = c: Cell | c.x % 2 == 0 witness Cell(0)

method P(b: bool, s: set<Cell>) returns (t: set<EvenCell>)
{
  if b {
    // in the following, every element selected by the set comprehension is an EvenCell value
    t := set c: EvenCell | c in s && c.x % 2 == 0;
  } else {
    t := set c: EvenCell | c in s;  // error: set elements are not constraint to satisfy EvenCell
  }
}
```

Both set comprehensions have type `set<EvenCell>`. However, in order for the latter to construct a set containing only `EvenCell` values, the evenness check needs to be performed somewhere. It cannot be generated by the compiler, because the subset constraint is a ghost expression. Therefore, (either set comprehensions with an element type that is a subset type must be banned altogether, or) the constraint must be checked by the verifier.

I think there's a way to change the type checker so that the verifier will (with no change) generate the appropriate checks. It is to insist that the type of a set comprehension is always a type `set<A>` where `A` is a base type (that is, not a subset type). If that were done for the program above, the assignment to `t` will be an assignment from `set<Cell>` to `set<EvenCell>`, and that will cause the verifier to generate the appropriate verification condition.

**Comment by RustanLeino** [comment]
A better suggestion (which is even easier to implement) is to insist that no bound variable of the set comprehension be of a subset type.

**Comment by robin-aws** [comment]
Copying over additional examples from #698 (which I'm closing as a duplicate of this one).

Verifies but crashes at runtime, because of the same issue in the context of runtime qualification:

```dafny
type Small = x: int | 0 <= x < 100 && x != 3

function method F(x: int): int
  requires x != 3
{
  if x == 3 then 1/0 else 100
}

method Main() {
  var b := forall n: Small :: F(n) == 100;
  print b, "\n";
}
```

Causes the Java and Go compilers to produce malformed code:

```dafny
newtype byte = x | -128 <= x < 128

method ByteIndexedArray(s: array<int>) returns (rev: array<int>)
  requires s.Length < 128
{
  rev := new int[s.Length];
  forall i: byte | 0 <= i < s.Length as byte {
    rev[i] := s[s.Length as byte - i - 1];
  }
}

method Main() {
  var s := new int[] [40, 41, 42, 43];
  var rev := ByteIndexedArray(s);
  print s[..], " ", rev[..], "\n";
}
```

**Comment by robin-aws** [comment]
Thanks @MikaelMayer, this is solution I had in mind as well. The title of the duplicate I closed (#698) also strongly implied that the fix would be to insert runtime evaluation of the predicate.

As long as we can produce clear errors messages as you describe, so that the rewriting is pure implementation detail, I'm very happy with this approach.

@RustanLeino is there a snag we're not seeing here that means this won't work, and we need to fall back to your more conservative solution(s)?

**Comment by RustanLeino** [comment]
Semantically, the solution proposed is good. That is, in any kind of comprehension or quantification, the types of the bound variables should be considered part of the range. I believe the verifier (but not the compiler) already does this.

So, we'd like to do the same for the compiler. I don't like the approach of doing a syntactic rewrite early in the pipeline, because we would then need to carry around information that remembers that parts of the range came from a syntactic desugaring. (For example, this would be needed in order to produce the error message mentioned above. It would also be needed to avoid duplicating well-formedness checks of the constraint, and it may also not give the verifier the flexibility to encode the type check in its own way.) Instead, I think we should change the compiler to emit the necessary code to test for the subset-type constraint.

To do this, we first need to make sure the constraint is compilable, that is, to make sure the type is _run-time checkable_. This is best done in the resolver and it's currently missing. I suggest we continue to let subset-type constraints be ghost, but that we perform an additional check on them to see if, by chance, they are compilable. For comprehensions in non-ghost contexts, the resolver would then forbid bound variables that have types that are not run-time checkable. For those comprehensions that live up to this restriction, the compiler then needs to be modified to check for the conditions at run time.

This mechanism can also be used to allow `is` tests for subset types.

**Comment by robin-aws** [comment]
It sounds like my `we can produce clear errors messages as you describe` precondition on my accepting Mikaël's solution evaluates to `false` (or at least requires a lot of complexity/extra work), in which case I agree with Rustan's solution. :)

> This mechanism can also be used to allow is tests for subset types.

Big +1 for this support! I want to make sure we focus on fixing the soundness bug ASAP first, but this would be a great follow-up change.

**Comment by robin-aws** [comment]
Also, although the rewriting approach may not be the ideal implementation, it DOES suggest a great way to test the soundness of the actual implementation: compare both solutions and ensure the behavior is always the same!

**Comment by MikaelMayer** [comment]
Ok, my PR #1522  is now ready for review....  but it has one performance problem.
I discovered that:

    type Class = x: Class? | x != null

and this is true for any declared class in Dafny, that is, every usual class type is a subset type.
Hence, for every comprehension, set, exists, map, forall, like this:;

    set c: Class | c in x && predicate(c)

even if `x` is another `set<Class>` and guaranteed to contain only non-null elements, the compiler will always emit the check that c != null, which is clearly not needed.
Plus, as said, it forces the constraint to be compilable even when it can be proved. It would be better if subset constraint stayed what they were supposed to be: *ghost*.

## The root problem and the root solution that will fix it all.

I believe I found the root of the problem and what we should do instead (the right fix, more precise)

In the following expression where `x` is of type `set<Cell>`

    set c: EvenCell :: c in x && doubleEvenCell(c) > 0;

the bounds of the variable of the set-comprehension are inferred to be of type set<Cell>. Since it is assumed that `c` will be any element of this bound, its **type should be assumed to be `Cell`**, and `EvenCell` needs to be asserted.
By "manually" fixing the type inference inside the Boogie generated file, it failed to verify, which is what I wanted. Moreover, I discovered that by replacing `EvenCell` by `Cell`, Dafny correctly emitted the verification condition for the requirement of `doubleEvenCell` as an assert instead of an assume,

I believe that I should be able to encode this transformation in the Dafny-to-Boogie translator.

### Benefits of this approach:

- **No more costly(?) implicit subset constraint check at run-time..**
- No more error stating that "X is a subset type and its constraint is not compilable" in a compiled context
- Developers can choose how they want to enforce the subset constraint.
- No need for a subset constraint to be compilable in any case.

### Problems of this approach:
- More verbose. It won't be possible to write `set c: EvenCell | c in x`, as it will complain that 'cannot prove that c is an EvenCell. Instead, programmers will have to explicitly write `set c: EvenCell | c in x && c.x % 2 == 0 in compiled contexts.

## What I'll do.

- Consider this issue as fixed once #1522  is merged.
- Open a new issue.
